Dockerfile

```
@@ -13,46 +13,307 @@
+RUN \
+if [[ "${APT_DEPS_IMAGE}" == "airflow-ci-apt-deps" ]]; then \
+    MINICLUSTER_URL=${MINICLUSTER_URL}${MINICLUSTER_VER}/minicluster-${MINICLUSTER_VER}-SNAPSHOT-bin.zip \
+    && MINICLUSTER_TMP_FILE=/tmp/minicluster.zip \
+    && curl -sL ${MINICLUSTER_URL} > ${MINICLUSTER_TMP_FILE} \
+    && unzip ${MINICLUSTER_TMP_FILE} -d /tmp \
+    && rm ${MINICLUSTER_TMP_FILE} \
+    ;\
+fi
+
+ENV PATH "$PATH:/tmp/hive/bin:$ADDITIONAL_PATH"
 
-FROM python:3.6-slim
-SHELL ["/bin/bash", "-xc"]
+############################################################################################################
+# This is the target image - it installs PIP and NPN dependencies including efficient caching
+# mechanisms - it might be used to build the bare airflow build or CI build
+# Parameters:
+#    APT_DEPS_IMAGE - image with APT dependencies. It might either be base deps image with airflow
+#                     dependencies or CI deps image that contains also CI-required dependencies
+############################################################################################################
+FROM ${APT_DEPS_IMAGE} as main
 
-ENV AIRFLOW_HOME=/usr/local/airflow
-ARG AIRFLOW_DEPS="all"
-ARG PYTHON_DEPS=""
-ARG BUILD_DEPS="freetds-dev libkrb5-dev libssl-dev libffi-dev libpq-dev git"
-ARG APT_DEPS="libsasl2-dev freetds-bin build-essential default-libmysqlclient-dev apt-utils curl rsync netcat locales"
+SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]
 
-ENV PATH="$HOME/.npm-packages/bin:$PATH"
+WORKDIR /opt/airflow
+
+RUN echo "Airflow version: ${AIRFLOW_VERSION}"
+
+ARG AIRFLOW_HOME=/opt/airflow
+ENV AIRFLOW_HOME=${AIRFLOW_HOME}
+
+RUN mkdir -pv ${AIRFLOW_HOME} \
+    && chown -R airflow.airflow ${AIRFLOW_HOME}
+
+# Increase the value here to force reinstalling Apache Airflow pip dependencies
+ENV FORCE_REINSTALL_ALL_PIP_DEPENDENCIES=1
+
+ARG CASS_DRIVER_BUILD_CONCURRENCY
+ARG CASS_DRIVER_NO_CYTHON
+ENV CASS_DRIVER_BUILD_CONCURRENCY=${CASS_DRIVER_BUILD_CONCURRENCY}
+ENV CASS_DRIVER_NO_CYTHON=${CASS_DRIVER_NO_CYTHON}
+
+# Airflow sources change frequently but dependency onfiguration won't change that often
+# We copy setup.py and other files needed to perform setup of dependencies
+# This way cache here will only be invalidated if any of the
+# version/setup configuration change but not when airflow sources change
+COPY --chown=airflow:airflow setup.py /opt/airflow/setup.py
+COPY --chown=airflow:airflow setup.cfg /opt/airflow/setup.cfg
+
+COPY --chown=airflow:airflow airflow/version.py /opt/airflow/airflow/version.py
+COPY --chown=airflow:airflow airflow/__init__.py /opt/airflow/airflow/__init__.py
+COPY --chown=airflow:airflow airflow/bin/airflow /opt/airflow/airflow/bin/airflow
+
+
+ARG AIRFLOW_EXTRAS="all"
+ENV AIRFLOW_EXTRAS=${AIRFLOW_EXTRAS}
+RUN echo "Installing with extras: ${AIRFLOW_EXTRAS}."
+
+ARG PIP_CACHE_DIRECTIVE
+ENV PIP_CACHE_DIRECTIVE=${PIP_CACHE_DIRECTIVE}
```

It's possible to set this directly via env variables https://pip.pypa.io/en/stable/user_guide/#environment-variables

```suggestion
ENV PIP_NO_CACHE_DIR=${PIP_NO_CACHE_DIR}
```

and then we don't need to add it in to the `pip` invocations

---

In this case actually we have to use the directive, because it can be more complex than --no-cache. When you build it during the Travis CI/DockerHub/Local Breeze build it will be set to use --cache-from=<DOWNLOADED_IMAGE> to use downloaded cache from DockerHub. It will only be set --no-cache when you choose to build it without cache (either via command line, setting the right environment variable before build or when you run CRON build on Travis).

---

Ah No.. It's pip cache :).... You are right... will be better this way.

---

Done.